modules/jfx.incubator.input/src/main/java/com/sun/jfx/incubator/scene/control/input/PHList.java

```diff
@@ -0,0 +1,225 @@
+        // don't expect many handlers, so linear search is ok
+        int sz = items.size();
+        boolean found = false;
+        for (int i = 0; i < sz; i++) {
+            Object x = items.get(i);
+            if (x instanceof EventHandlerPriority p) {
+                if (p.priority == priority.priority) {
+                    found = true;
+                    continue;
+                } else if (p.priority < priority.priority) {
+                    return found ? i : -(i + 1);
+                }
+            }
+        }
+        return found ? sz : -(sz + 1);
+    }
+
+    /**
+     * A client interface for the {@link #forEach(Client)} method.
+     * @param <T> the event type
+     */
+    @FunctionalInterface
+    public static interface Client<T extends Event> {
+        /**
+         * This method gets called for each handler in the order of priority.
+         * The client may signal to stop iterating by returning false from this method.
+         *
+         * @param pri the priority
+         * @param h the handler (can be null)
+         * @return true to continue the process, false to stop
+         */
+        public boolean accept(EventHandlerPriority pri, EventHandler<T> h);
+    }
+
+    /**
+     * Invokes the {@code client} for each handler in the order of priority.
+     * @param <T> the event type
+     * @param client the client reference
+     */
+    public <T extends Event> void forEach(Client<T> client) {
+        EventHandlerPriority pri = null;
+        boolean stop;
+        int sz = items.size();
+        for (int i = 0; i < sz; i++) {
+            Object x = items.get(i);
+            if (x instanceof EventHandlerPriority p) {
+                pri = p;
+                if (isNullOrPriority(i + 1)) {
+                    stop = !client.accept(pri, null);
+                } else {
+                    continue;
+                }
+            } else {
+                // it's a handler, cannot be null
+                stop = !client.accept(pri, (EventHandler<T>)x);
+            }
+            if (stop) {
+                break;
+            }
+        }
+    }
+
+    /**
+     * Removes all the entries with the specified priorities.
+     * @return true if list is empty as a result
+     */
+    public boolean removeHandlers(Set<EventHandlerPriority> priorities) {
+        boolean remove = false;
+        for (int i = 0; i < items.size();) {
+            Object x = items.get(i);
+            if (x instanceof EventHandlerPriority p) {
+                if (priorities.contains(p)) {
+                    remove = true;
+                    items.remove(i);
+                } else {
+                    remove = false;
+                    i++;
+                }
+            } else {
+                if (remove) {
+                    items.remove(i);
+                } else {
+                    i++;
+                }
+            }
+        }
+        return items.size() == 0;
```

Comment:
Use of ArrayList.isEmpty() would look nicer.
return items.size() == 0; -> return items.isEmpty();